api/src/main/java/org/apache/iceberg/io/CloseableIterable.java

```diff
@@ -73,6 +74,36 @@ public CloseableIterator<E> iterator() {
     };
   }
 
+  /**
+   * Will run the given runnable when {@link CloseableIterable#close()} has been called.
+   *
+   * @param iterable The underlying {@link CloseableIterable} to iterate over
+   * @param onCompletionRunnable The runnable to run after the underlying iterable was closed
+   * @param <E> The type of the underlying iterable
+   * @return A new {@link CloseableIterable} where the runnable will be executed as the final step
+   *     after {@link CloseableIterable#close()} has been called
+   */
+  static <E> CloseableIterable<E> whenComplete(
+      CloseableIterable<E> iterable, Runnable onCompletionRunnable) {
+    Preconditions.checkNotNull(
+        onCompletionRunnable, "Cannot execute a null Runnable after completion");
```

Comment:
Nit: Maybe `Invalid runnable: null`? I usually figure that the stack trace will have the remaining information that'd needed. In that it will have `whenComplete` in it, etc.

Reply:
good point, I'll follow up on this, since the code was merged to master and I just rebased this PR

Reply:
addressed in https://github.com/apache/iceberg/pull/5433